{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "from bertopic import BERTopic\n",
    "import datetime\n",
    "import os \n",
    "import numpy as np"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Basic BertTopic"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Functions/definitons"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 211,
   "metadata": {},
   "outputs": [],
   "source": [
    "# PATHS\n",
    "train_data_path = '../data/titles_and_abstracts_processed_train.csv'\n",
    "test_data_path = '../data/titles_and_abstracts_processed_test.csv'\n",
    "models_path = '../models/' \n",
    "results_path = '../results/bertopic'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 212,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_now_str():\n",
    "    return datetime.datetime.now().strftime(\"%Y-%m-%d_%H-%M-%S\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 213,
   "metadata": {},
   "outputs": [],
   "source": [
    "def transform_strings_to_arrays(df, col_names = ['tokenized_sentences', 'tokenized_words', 'tokenized_words_processed']):\n",
    "    for col in col_names:\n",
    "        df[col] = df[col].apply(eval)\n",
    "    return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 214,
   "metadata": {},
   "outputs": [],
   "source": [
    "# basic BertTopic keyword extraction\n",
    "def train_transform_save(train_data, model_save_name):\n",
    "    \n",
    "    # train transform\n",
    "    topic_model = BERTopic()\n",
    "    topics, probs = topic_model.fit_transform(train_data.values)\n",
    "\n",
    "    # save model\n",
    "    topic_model.save(model_save_name)\n",
    "\n",
    "    return topic_model, topics, probs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 228,
   "metadata": {},
   "outputs": [],
   "source": [
    "def load_transform_save(data, model_save_name, results_path):\n",
    "\n",
    "    # load model\n",
    "    loaded_model = BERTopic.load(model_save_name)\n",
    "\n",
    "    # transform for data \n",
    "    samples_topics, samples_probs = loaded_model.transform(data.values)\n",
    "    res_df = pd.DataFrame({\n",
    "        'PMID': np.unique(data.index),\n",
    "        'topic_number': samples_topics,\n",
    "        'topic_probs': samples_probs,\n",
    "        \"topic_keywords\": [loaded_model.get_topic(topic_number) for topic_number in samples_topics]\n",
    "    })\n",
    "    res_df.to_csv(results_path)\n",
    "    return loaded_model, res_df\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Processed data (with steming and stop words removal)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 229,
   "metadata": {},
   "outputs": [],
   "source": [
    "train_full_data = transform_strings_to_arrays(pd.read_csv(train_data_path))\n",
    "test_full_data = transform_strings_to_arrays(pd.read_csv(test_data_path))\n",
    "\n",
    "train_data = train_full_data.groupby(by = ['PMID'])['tokenized_words_processed'].agg(lambda x: ' '.join(x.values[0] + x.values[1]))\n",
    "test_data = test_full_data.groupby(by = ['PMID'])['tokenized_words_processed'].agg(lambda x: ' '.join(x.values[0] + x.values[1]))\n",
    "\n",
    "model_name = f'berttopic_processed_data_{get_now_str()}'\n",
    "model_save_name = os.path.join(models_path, model_name)\n",
    "result_path_train = os.path.join(results_path, f'{model_name}_train.csv')\n",
    "result_path_test = os.path.join(results_path, f'{model_name}_test.csv')\n",
    "\n",
    "topic_model, topics, probs = train_transform_save(train_data, model_save_name)\n",
    "_, res_df_train = load_transform_save(train_data, model_save_name, result_path_train)\n",
    "_, res_df_test = load_transform_save(test_data, model_save_name, result_path_test)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Processed data example"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 230,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'dctn4 modifi chronic pseudomona aeruginosa infect cystic fibrosi pseudomona aeruginosa infect cystic fibrosi cf patient associ wors pulmonari diseas shorter surviv chronic infect cpa associ reduc lung function faster rate lung declin increas rate exacerb shorter surviv exom sequenc extrem phenotyp design recent shown isoform dynactin 4 dctn4 influenc infect cf lead wors respiratori diseas purpos studi investig role dctn4 missens variant infect incid age infect chronic infect incid cohort adult cf patient singl centr polymeras chain reaction direct sequenc screen dna sampl dctn4 variant total 121 adult cf patient cochin hospit cf centr includ carri cftr defect 103 develop 1 pulmonari infect 68 patient cpa dctn4 variant identifi 24 cf patient infect 17 cf patient infect patient cpa 29 dctn4 missens variant vs 23 patient cpa interestingli tend frequent observ cf patient cpa patient cpa vs dctn4 missens variant tend frequent male cf patient cpa bear class mutat male cf patient cpa bear class mutat p observ reinforc dctn4 missens variant especi involv pathogenesi cpa male cf'"
      ]
     },
     "execution_count": 230,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "train_data.iloc[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 231,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Topic</th>\n",
       "      <th>Count</th>\n",
       "      <th>Name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>-1</td>\n",
       "      <td>1421</td>\n",
       "      <td>-1_patient_studi_effect_associ</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0</td>\n",
       "      <td>124</td>\n",
       "      <td>0_neuron_brain_cell_express</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>120</td>\n",
       "      <td>1_cell_cancer_tumor_express</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2</td>\n",
       "      <td>115</td>\n",
       "      <td>2_nanoparticl_surfac_coat_properti</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>3</td>\n",
       "      <td>100</td>\n",
       "      <td>3_speci_forest_popul_bird</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>4</td>\n",
       "      <td>100</td>\n",
       "      <td>4_fractur_knee_bone_spine</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>5</td>\n",
       "      <td>94</td>\n",
       "      <td>5_health_student_practic_profession</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>6</td>\n",
       "      <td>61</td>\n",
       "      <td>6_protein_bind_structur_ligand</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>7</td>\n",
       "      <td>58</td>\n",
       "      <td>7_ventricular_cardiac_patient_lv</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>8</td>\n",
       "      <td>58</td>\n",
       "      <td>8_cancer_surviv_patient_lung</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>9</td>\n",
       "      <td>52</td>\n",
       "      <td>9_eye_thick_optic_macular</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>10</td>\n",
       "      <td>50</td>\n",
       "      <td>10_parent_depress_child_children</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>11</td>\n",
       "      <td>47</td>\n",
       "      <td>11_strain_acid_product_ferment</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>12</td>\n",
       "      <td>45</td>\n",
       "      <td>12_brain_cognit_cortic_connect</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>13</td>\n",
       "      <td>44</td>\n",
       "      <td>13_movement_speed_speech_walk</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>14</td>\n",
       "      <td>42</td>\n",
       "      <td>14_pregnanc_matern_women_birth</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>15</td>\n",
       "      <td>42</td>\n",
       "      <td>15_transplant_kidney_renal_liver</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>16</td>\n",
       "      <td>41</td>\n",
       "      <td>16_lipid_apo_cholesterol_liver</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>17</td>\n",
       "      <td>40</td>\n",
       "      <td>17_teeth_dental_restor_periodont</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>18</td>\n",
       "      <td>40</td>\n",
       "      <td>18_viru_virus_viral_infect</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>19</td>\n",
       "      <td>39</td>\n",
       "      <td>19_plant_gene_arabidopsi_root</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>20</td>\n",
       "      <td>38</td>\n",
       "      <td>20_gene_methyl_dna_bind</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>21</td>\n",
       "      <td>35</td>\n",
       "      <td>21_breast_cancer_patient_tumor</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>22</td>\n",
       "      <td>34</td>\n",
       "      <td>22_diabet_hba1c_type_insulin</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>23</td>\n",
       "      <td>33</td>\n",
       "      <td>23_insulin_diabet_adipos_glucos</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>24</td>\n",
       "      <td>32</td>\n",
       "      <td>24_obes_food_weight_children</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>25</td>\n",
       "      <td>31</td>\n",
       "      <td>25_bone_osteogen_osteoblast_chondrocyt</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>26</td>\n",
       "      <td>30</td>\n",
       "      <td>26_skin_psoriasi_patient_ra</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>27</td>\n",
       "      <td>28</td>\n",
       "      <td>27_isol_strain_infect_resist</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>28</td>\n",
       "      <td>28</td>\n",
       "      <td>28_cardiomyocyt_cardiac_myocardi_hypertrophi</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>29</td>\n",
       "      <td>24</td>\n",
       "      <td>29_coumarin_compound_extract_metabolit</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>30</td>\n",
       "      <td>22</td>\n",
       "      <td>30_schizophrenia_asenapin_bid_antipsychot</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>31</td>\n",
       "      <td>22</td>\n",
       "      <td>31_asthma_inflamm_lung_inhal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>32</td>\n",
       "      <td>21</td>\n",
       "      <td>32_gene_lncrna_cell_edit</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>33</td>\n",
       "      <td>20</td>\n",
       "      <td>33_aneurysm_arteri_embol_effus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>34</td>\n",
       "      <td>20</td>\n",
       "      <td>34_bowel_hp_gastrostomi_diseas</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>35</td>\n",
       "      <td>19</td>\n",
       "      <td>35_thyroid_hypothyroid_hormon_case</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>36</td>\n",
       "      <td>19</td>\n",
       "      <td>36_coli_strain_bacteri_colistin</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>37</td>\n",
       "      <td>19</td>\n",
       "      <td>37_cardiovascular_risk_cvd_age</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>38</td>\n",
       "      <td>18</td>\n",
       "      <td>38_pain_tofacitinib_plll_treatment</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>39</td>\n",
       "      <td>18</td>\n",
       "      <td>39_neutrophil_cd177_th17_platelet</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>40</td>\n",
       "      <td>17</td>\n",
       "      <td>40_soil_plant_herbicid_o3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>41</td>\n",
       "      <td>17</td>\n",
       "      <td>41_tumor_oscc_ctc_cancer</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>42</td>\n",
       "      <td>17</td>\n",
       "      <td>42_mirna_microrna_express_mrna</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>43</td>\n",
       "      <td>16</td>\n",
       "      <td>43_vaccin_serotyp_pertussi_pcv13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45</th>\n",
       "      <td>44</td>\n",
       "      <td>16</td>\n",
       "      <td>44_cognit_dementia_phish_mood</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46</th>\n",
       "      <td>45</td>\n",
       "      <td>16</td>\n",
       "      <td>45_hypertens_bp_combin_felodipin</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47</th>\n",
       "      <td>46</td>\n",
       "      <td>16</td>\n",
       "      <td>46_water_sediment_pcdd_fs</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48</th>\n",
       "      <td>47</td>\n",
       "      <td>15</td>\n",
       "      <td>47_bladder_biopsi_cancer_prostat</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>49</th>\n",
       "      <td>48</td>\n",
       "      <td>15</td>\n",
       "      <td>48_smoke_cigarett_nicotin_smoker</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50</th>\n",
       "      <td>49</td>\n",
       "      <td>14</td>\n",
       "      <td>49_mitochondri_mitochondria_lg_mutat</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51</th>\n",
       "      <td>50</td>\n",
       "      <td>14</td>\n",
       "      <td>50_malaria_parasit_falciparum_plasmodium</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>52</th>\n",
       "      <td>51</td>\n",
       "      <td>13</td>\n",
       "      <td>51_medicin_reliabl_tb_rsrr</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53</th>\n",
       "      <td>52</td>\n",
       "      <td>13</td>\n",
       "      <td>52_pneumonia_antimicrobi_isol_hmkp</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54</th>\n",
       "      <td>53</td>\n",
       "      <td>13</td>\n",
       "      <td>53_hiv_art_saliva_infect</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55</th>\n",
       "      <td>54</td>\n",
       "      <td>13</td>\n",
       "      <td>54_stent_ureter_renal_strictur</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56</th>\n",
       "      <td>55</td>\n",
       "      <td>12</td>\n",
       "      <td>55_hpv_women_chlamydia_infect</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57</th>\n",
       "      <td>56</td>\n",
       "      <td>11</td>\n",
       "      <td>56_pd_mog_parkinson_ataxia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58</th>\n",
       "      <td>57</td>\n",
       "      <td>11</td>\n",
       "      <td>57_zinc_cell_apoptosi_nonylphenol</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59</th>\n",
       "      <td>58</td>\n",
       "      <td>10</td>\n",
       "      <td>58_coronari_arteri_bypass_collater</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    Topic  Count                                          Name\n",
       "0      -1   1421                -1_patient_studi_effect_associ\n",
       "1       0    124                   0_neuron_brain_cell_express\n",
       "2       1    120                   1_cell_cancer_tumor_express\n",
       "3       2    115            2_nanoparticl_surfac_coat_properti\n",
       "4       3    100                     3_speci_forest_popul_bird\n",
       "5       4    100                     4_fractur_knee_bone_spine\n",
       "6       5     94           5_health_student_practic_profession\n",
       "7       6     61                6_protein_bind_structur_ligand\n",
       "8       7     58              7_ventricular_cardiac_patient_lv\n",
       "9       8     58                  8_cancer_surviv_patient_lung\n",
       "10      9     52                     9_eye_thick_optic_macular\n",
       "11     10     50              10_parent_depress_child_children\n",
       "12     11     47                11_strain_acid_product_ferment\n",
       "13     12     45                12_brain_cognit_cortic_connect\n",
       "14     13     44                 13_movement_speed_speech_walk\n",
       "15     14     42                14_pregnanc_matern_women_birth\n",
       "16     15     42              15_transplant_kidney_renal_liver\n",
       "17     16     41                16_lipid_apo_cholesterol_liver\n",
       "18     17     40              17_teeth_dental_restor_periodont\n",
       "19     18     40                    18_viru_virus_viral_infect\n",
       "20     19     39                 19_plant_gene_arabidopsi_root\n",
       "21     20     38                       20_gene_methyl_dna_bind\n",
       "22     21     35                21_breast_cancer_patient_tumor\n",
       "23     22     34                  22_diabet_hba1c_type_insulin\n",
       "24     23     33               23_insulin_diabet_adipos_glucos\n",
       "25     24     32                  24_obes_food_weight_children\n",
       "26     25     31        25_bone_osteogen_osteoblast_chondrocyt\n",
       "27     26     30                   26_skin_psoriasi_patient_ra\n",
       "28     27     28                  27_isol_strain_infect_resist\n",
       "29     28     28  28_cardiomyocyt_cardiac_myocardi_hypertrophi\n",
       "30     29     24        29_coumarin_compound_extract_metabolit\n",
       "31     30     22     30_schizophrenia_asenapin_bid_antipsychot\n",
       "32     31     22                  31_asthma_inflamm_lung_inhal\n",
       "33     32     21                      32_gene_lncrna_cell_edit\n",
       "34     33     20                33_aneurysm_arteri_embol_effus\n",
       "35     34     20                34_bowel_hp_gastrostomi_diseas\n",
       "36     35     19            35_thyroid_hypothyroid_hormon_case\n",
       "37     36     19               36_coli_strain_bacteri_colistin\n",
       "38     37     19                37_cardiovascular_risk_cvd_age\n",
       "39     38     18            38_pain_tofacitinib_plll_treatment\n",
       "40     39     18             39_neutrophil_cd177_th17_platelet\n",
       "41     40     17                     40_soil_plant_herbicid_o3\n",
       "42     41     17                      41_tumor_oscc_ctc_cancer\n",
       "43     42     17                42_mirna_microrna_express_mrna\n",
       "44     43     16              43_vaccin_serotyp_pertussi_pcv13\n",
       "45     44     16                 44_cognit_dementia_phish_mood\n",
       "46     45     16              45_hypertens_bp_combin_felodipin\n",
       "47     46     16                     46_water_sediment_pcdd_fs\n",
       "48     47     15              47_bladder_biopsi_cancer_prostat\n",
       "49     48     15              48_smoke_cigarett_nicotin_smoker\n",
       "50     49     14          49_mitochondri_mitochondria_lg_mutat\n",
       "51     50     14      50_malaria_parasit_falciparum_plasmodium\n",
       "52     51     13                    51_medicin_reliabl_tb_rsrr\n",
       "53     52     13            52_pneumonia_antimicrobi_isol_hmkp\n",
       "54     53     13                      53_hiv_art_saliva_infect\n",
       "55     54     13                54_stent_ureter_renal_strictur\n",
       "56     55     12                 55_hpv_women_chlamydia_infect\n",
       "57     56     11                    56_pd_mog_parkinson_ataxia\n",
       "58     57     11             57_zinc_cell_apoptosi_nonylphenol\n",
       "59     58     10            58_coronari_arteri_bypass_collater"
      ]
     },
     "execution_count": 231,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "topic_model.get_topic_info()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 241,
   "metadata": {},
   "outputs": [],
   "source": [
    "# example \n",
    "i = 11\n",
    "doc = train_data.iloc[i]\n",
    "doc_topic = topics[i]\n",
    "doc_prob = probs[i]\n",
    "topic_info = topic_model.get_topic_info(doc_topic)\n",
    "topic_words = topic_model.get_topic(doc_topic)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 243,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'neutral forbidden link morpholog match assembl mutualist hawkmoth plant network major challeng evolutionari ecolog understand process shape pattern interact speci commun level pollin flower long corolla tube hawkmoth invok showcas model recent optim forag model predict close associ mouthpart length corolla depth visit flower favour trait converg special commun level assess hawkmoth frequent pollin plant floral tube length similar probosci length morpholog match hypothesi abund process neutral hypothesi ecolog trait mismatch constraint forbidden link hypothesi process structur hawkmoth plant mutualist network commun biogeograph region south america found converg morpholog trait commun distribut morpholog differ hawkmoth plant consist expect morpholog match hypothesi commun remain commun ecoton distinct biogeograph area interact predict neutral hypothesi find consist idea diffus drive evolut extrem long proboscis flower tube highlight import morpholog trait forbidden link hypothesi structur interact mutualist partner reveal role process complex previous'"
      ]
     },
     "execution_count": 243,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "doc"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 244,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Topic</th>\n",
       "      <th>Count</th>\n",
       "      <th>Name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>3</td>\n",
       "      <td>100</td>\n",
       "      <td>3_speci_forest_popul_bird</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Topic  Count                       Name\n",
       "0      3    100  3_speci_forest_popul_bird"
      ]
     },
     "execution_count": 244,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "topic_info"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 246,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[('speci', 0.04541899474237719),\n",
       " ('forest', 0.01986315689650549),\n",
       " ('popul', 0.01879033701326042),\n",
       " ('bird', 0.0155259175762611),\n",
       " ('climat', 0.01385799346446381),\n",
       " ('plant', 0.013240451447144554),\n",
       " ('divers', 0.013054858242023333),\n",
       " ('genet', 0.011344986058897222),\n",
       " ('tree', 0.01119505679918582),\n",
       " ('phylogenet', 0.0104014699784837)]"
      ]
     },
     "execution_count": 246,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "topic_words"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Raw data (without stemming and stopwords removal)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 247,
   "metadata": {},
   "outputs": [],
   "source": [
    "train_full_data = transform_strings_to_arrays(pd.read_csv(train_data_path))\n",
    "test_full_data = transform_strings_to_arrays(pd.read_csv(test_data_path))\n",
    "\n",
    "train_data = train_full_data.groupby(by = ['PMID'])['Content'].agg(lambda x: f'{x.values[0]}. {x.values[1]}')\n",
    "test_data = test_full_data.groupby(by = ['PMID'])['Content'].agg(lambda x: f'{x.values[0]}. {x.values[1]}')\n",
    "\n",
    "model_name = f'bertopic_raw_data_{get_now_str()}'\n",
    "model_save_name = os.path.join(models_path, model_name)\n",
    "result_path_train = os.path.join(results_path, f'{model_name}_train.csv')\n",
    "result_path_test = os.path.join(results_path, f'{model_name}_test.csv')\n",
    "\n",
    "topic_model, topics, probs = train_transform_save(train_data, model_save_name)\n",
    "_, res_df_train = load_transform_save(train_data, model_save_name, result_path_train)\n",
    "_, res_df_test = load_transform_save(test_data, model_save_name, result_path_test)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Raw data example"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 249,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'DCTN4 as a modifier of chronic Pseudomonas aeruginosa infection in cystic fibrosis. Pseudomonas aeruginosa (Pa) infection in cystic fibrosis (CF) patients is associated with worse long-term pulmonary disease and shorter survival, and chronic Pa infection (CPA) is associated with reduced lung function, faster rate of lung decline, increased rates of exacerbations and shorter survival. By using exome sequencing and extreme phenotype design, it was recently shown that isoforms of dynactin 4 (DCTN4) may influence Pa infection in CF, leading to worse respiratory disease. The purpose of this study was to investigate the role of DCTN4 missense variants on Pa infection incidence, age at first Pa infection and chronic Pa infection incidence in a cohort of adult CF patients from a single centre. Polymerase chain reaction and direct sequencing were used to screen DNA samples for DCTN4 variants. A total of 121 adult CF patients from the Cochin Hospital CF centre have been included, all of them carrying two CFTR defects: 103 developed at least 1 pulmonary infection with Pa, and 68 patients of them had CPA. DCTN4 variants were identified in 24% (29/121) CF patients with Pa infection and in only 17% (3/18) CF patients with no Pa infection. Of the patients with CPA, 29% (20/68) had DCTN4 missense variants vs 23% (8/35) in patients without CPA. Interestingly, p.Tyr263Cys tend to be more frequently observed in CF patients with CPA than in patients without CPA (4/68 vs 0/35), and DCTN4 missense variants tend to be more frequent in male CF patients with CPA bearing two class II mutations than in male CF patients without CPA bearing two class II mutations (P = 0.06). Our observations reinforce that DCTN4 missense variants, especially p.Tyr263Cys, may be involved in the pathogenesis of CPA in male CF.'"
      ]
     },
     "execution_count": 249,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "train_data.iloc[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 250,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Topic</th>\n",
       "      <th>Count</th>\n",
       "      <th>Name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>-1</td>\n",
       "      <td>1432</td>\n",
       "      <td>-1_the_of_and_in</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0</td>\n",
       "      <td>168</td>\n",
       "      <td>0_care_health_and_to</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>106</td>\n",
       "      <td>1_brain_to_the_in</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2</td>\n",
       "      <td>82</td>\n",
       "      <td>2_production_soil_the_of</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>3</td>\n",
       "      <td>78</td>\n",
       "      <td>3_cells_that_the_in</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60</th>\n",
       "      <td>59</td>\n",
       "      <td>13</td>\n",
       "      <td>59_exercise_muscle_muscles_dystrophy</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>61</th>\n",
       "      <td>60</td>\n",
       "      <td>13</td>\n",
       "      <td>60_dpd_parks_spatial_travel</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>62</th>\n",
       "      <td>61</td>\n",
       "      <td>13</td>\n",
       "      <td>61_thyroid_tc_cancer_95</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>63</th>\n",
       "      <td>62</td>\n",
       "      <td>12</td>\n",
       "      <td>62_schizophrenia_asenapine_antipsychotic_bid</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>64</th>\n",
       "      <td>63</td>\n",
       "      <td>10</td>\n",
       "      <td>63_datasets_ppi_modules_fusions</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>65 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    Topic  Count                                          Name\n",
       "0      -1   1432                              -1_the_of_and_in\n",
       "1       0    168                          0_care_health_and_to\n",
       "2       1    106                             1_brain_to_the_in\n",
       "3       2     82                      2_production_soil_the_of\n",
       "4       3     78                           3_cells_that_the_in\n",
       "..    ...    ...                                           ...\n",
       "60     59     13          59_exercise_muscle_muscles_dystrophy\n",
       "61     60     13                   60_dpd_parks_spatial_travel\n",
       "62     61     13                       61_thyroid_tc_cancer_95\n",
       "63     62     12  62_schizophrenia_asenapine_antipsychotic_bid\n",
       "64     63     10               63_datasets_ppi_modules_fusions\n",
       "\n",
       "[65 rows x 3 columns]"
      ]
     },
     "execution_count": 250,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "topic_model.get_topic_info()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 251,
   "metadata": {},
   "outputs": [],
   "source": [
    "# example \n",
    "i = 11\n",
    "doc = train_data.iloc[i]\n",
    "doc_topic = topics[i]\n",
    "doc_prob = probs[i]\n",
    "topic_info = topic_model.get_topic_info(doc_topic)\n",
    "topic_words = topic_model.get_topic(doc_topic)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 252,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "\"Beyond neutral and forbidden links: morphological matches and the assembly of mutualistic hawkmoth - plant networks. A major challenge in evolutionary ecology is to understand how co-evolutionary processes shape patterns of interactions between species at community level. Pollination of flowers with long corolla tubes by long-tongued hawkmoths has been invoked as a showcase model of co-evolution. Recently, optimal foraging models have predicted that there might be a close association between mouthparts' length and the corolla depth of the visited flowers, thus favouring trait convergence and specialization at community level. Here, we assessed whether hawkmoths more frequently pollinate plants with floral tube lengths similar to their proboscis lengths (morphological match hypothesis) against abundance -based processes (neutral hypothesis) and ecological trait mismatches constraints (forbidden links hypothesis), and how these processes structure hawkmoth - plant mutualistic networks from five communities in four biogeographical regions of South America. We found convergence in morphological traits across the five communities and that the distribution of morphological differences between hawkmoths and plants is consistent with expectations under the morphological match hypothesis in three of the five communities. In the two remaining communities, which are ecotones between two distinct biogeographical areas, interactions are better predicted by the neutral hypothesis. Our findings are consistent with the idea that diffuse co-evolution drives the evolution of extremely long proboscises and flower tubes, and highlight the importance of morphological traits, beyond the forbidden links hypothesis, in structuring interactions between mutualistic partners, revealing that the role of niche-based processes can be much more complex than previously known.\""
      ]
     },
     "execution_count": 252,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "doc"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 253,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Topic</th>\n",
       "      <th>Count</th>\n",
       "      <th>Name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>4</td>\n",
       "      <td>70</td>\n",
       "      <td>4_species_the_of_and</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Topic  Count                  Name\n",
       "0      4     70  4_species_the_of_and"
      ]
     },
     "execution_count": 253,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "topic_info"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 254,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[('species', 0.040162911115070137),\n",
       " ('the', 0.016339623447882926),\n",
       " ('of', 0.014235148640351948),\n",
       " ('and', 0.012700373298775505),\n",
       " ('we', 0.01182749835321019),\n",
       " ('forests', 0.01173852168148737),\n",
       " ('in', 0.011656504596709786),\n",
       " ('that', 0.011253687294829333),\n",
       " ('to', 0.011098637279477231),\n",
       " ('climate', 0.010928281659616299)]"
      ]
     },
     "execution_count": 254,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "topic_words"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3.9.13 ('nlp')",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.13"
  },
  "orig_nbformat": 4,
  "vscode": {
   "interpreter": {
    "hash": "88ba28277a1630fc5f744f385ba08b8a5477e64ab64787e3b7b12637e39579c4"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
